src/messaging/ChannelContext.cpp

```diff
@@ -0,0 +1,407 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <messaging/Channel.h>
+#include <messaging/ChannelContext.h>
+#include <messaging/ExchangeMgr.h>
+
+namespace chip {
+namespace Messaging {
+
+void ChannelContextDeletor::Release(ChannelContext * context)
+{
+    context->mExchangeManager->ReleaseChannelContext(context);
+}
+
+void ChannelContext::Start(const ChannelBuilder & builder)
+{
+    if (mState != ChannelState::kChanneState_None)
+        return;
+    EnterPreparingState(builder);
+}
+
+ExchangeContext * ChannelContext::NewExchange(ExchangeDelegate * delegate)
+{
+    if (GetState() != ChannelState::kChanneState_Ready)
+        return nullptr;
```

Comment:
Yes. Consider the following scenario:

1. The channel is ready.
2. The channel is broken immediate due to some unknown reason.
3. The application gets the ready event callback
4. The application calls `ChannelContext::NewExchange` inside the callback function.

Reply:
Why can't you create an exchange with a broken channel?

Reply:
What I mean is, clearly the channel can break after NewExchange - it can fail at any time. Applications will have to be able to deal with broken exchanges.

I'm not sure why it breaking before the callback in your sequence should have a different result than it breaking after. It seems like both could result in a broken exchange, rather than a null pointer.

It feels wrong to me to bail on a New() function except in the following ways:

- success
- assert due to precondition fail
- out of memory/resource error
 
If the application will known the channel failed, we can make it a precondition not to call NewExchange and just assert.

If it doesn't know, we can proceed with creating the exchange and then inform the application that the exchange is broken.

What feels wrong - just throw nullptrs at the application, out of the blue. It's a recipe for null pointer errors.

Reply:
I'll make it a precondition using assert and don't return null here. But further in the `ExchangeManager::AllocContext` function, it may returns null is not enough resource for exchange, so the application still need to check nullptr.

I'm not sure returning a broken exchange is a good idea. because a healthy exchange will trigger a fail event when it becomes invalid. But a broken exchange will never trigger an event.